File: Network_Construction/ROI2graphCorrelation.m
function [ graphWithROIs ] = ROI2graphCorrelation(imageWithROI)
	%	generate graph from given ROI 
	%   input: a ROI matrix where NaN indicates non-ROI region, 
	%   output: a graph adjacent matrix 
	%	Each node in the graph is a pixel in the ROI, and the
	%   node connect to its adjacent 26 pixels, the weight of each edge is the intensity
	%   difference between adjacent pixels

	%	imageWithROI: ROI matrix
	%	graph:adjacent matrix of a graph
	% I = load('./DataNew/imageWithROIs.mat');
	seq = [-2,-1,0,1,2];
	mask = zeros(125,3);
	for i = 0:124
		a = rem(i,5) + 1;
		b = floor((i - floor(i / 25)*25)/5) + 1;
		c = floor(i/25) + 1;
		mask(i+1,:) = [seq(c),seq(b),seq(a)];
	end

    [width,height,depth]=size(imageWithROI);

    map=zeros(width,height,depth);
    nNodes=0;
    for i=1:width
        for j=1:height
            for k=1:depth
                if (~isnan(imageWithROI(i,j,k)) && (imageWithROI(i,j,k)~=0))
                    nNodes=nNodes+1;
                    map(i,j,k)=nNodes;
                end
            end
        end
	end
	
	valueNearby = ConstructNeighbor(imageWithROI,map);
	%disp(valueNearby)
    graphWithROIs=zeros(nNodes,nNodes);
    for i=1:width
        for j=1:height
            for k=1:depth
                if (map(i,j,k)~=0)
                    node0=map(i,j,k);
					value0=valueNearby{i,j,k};
					for ii=1:width
						for jj=1:height
							for kk=1:depth
								if (map(ii,jj,kk)>node0)
									node1=map(ii,jj,kk);
									value1=valueNearby{ii,jj,kk};
                                    %disp('value0')
									%disp(value0)
                                    %disp('value1')
									%disp(value1)
									correlation = ComputeNormalizedCorrelation(value0,value1);
									graphWithROIs(int16(node0),int16(node1)) = correlation;  
									graphWithROIs(int16(node1),int16(node0)) = correlation; 
								end
							end
						end
					end
                end
            end
		end
	end
end

File: Network_Construction/ROI2graphCorrelation.py
import numpy as np
from ConstructNeighbor import ConstructNeighbor
from ComputeNormalizedCorrelation import ComputeNormalizedCorrelation
import math

def ROI2graphCorrelation(imageWithROI):
    seq = [-2,-1,0,1,2]
    mask = np.zeros([125,3])
    for i in range(0, 125):
    	a = int(i % 5)
    	b = int(math.floor((i - math.floor(i / 25)*25)/5))
    	c = int(math.floor(i/25))
    	#mask(i+1,:) = [seq(c),seq(b),seq(a)]
        mask[i,0] = seq[c]
        mask[i,1] = seq[b]
        mask[i,2] = seq[a]
    #print(imageWithROI)
    width = imageWithROI.shape[0]
    height = imageWithROI.shape[1]
    depth = imageWithROI.shape[2]
    
    #print('w {} h {} d {}'.format(width, height, depth))
    map=np.zeros([width,height,depth])
    nNodes=0
    for i in range(1,width+1):
        for j in range(1, height+1):
            for k in range(1,depth+1):
                if (~np.isnan(imageWithROI[i-1,j-1,k-1]) and (not imageWithROI[i-1,j-1,k-1]==0)):
                    nNodes=nNodes+1
                    map[i-1,j-1,k-1]=nNodes
	
	valueNearby = ConstructNeighbor(imageWithROI,map)
    #print(valueNearby)
	
    graphWithROIs=np.zeros([nNodes,nNodes])
    for i in range(1,width+1):
        for j in range(1,height+1):
            for k in range(1,depth+1):
                if (not map[i-1,j-1,k-1]==0):
                    node0=map[i-1,j-1,k-1]
                    value0=valueNearby[i-1,j-1,k-1]
                    for ii in range(1,width+1):
                        for jj in range(1,height+1):
                            for kk in range(1,depth+1):
                                if (map[ii-1,jj-1,kk-1]>node0):
                                    node1=map[ii-1,jj-1,kk-1]
                                    value1=valueNearby[ii-1,jj-1,kk-1]
                                    #print('v0 {} v1 {}'.format(value0, value1))
                                    correlation = ComputeNormalizedCorrelation(value0,value1)
                                    #print('graphwithrois size'.format(graphWithROIs.shape))
                                    #int16->int >>> TODO check loss here
                                    graphWithROIs[int(node0-1),int(node1-1)]= correlation  
                                    graphWithROIs[int(node1-1),int(node0-1)] = correlation 


    return graphWithROIs
